http-server-netty/src/main/java/io/micronaut/http/server/netty/multipart/NettyStreamingFileUpload.java

```diff
@@ -114,29 +114,42 @@ public boolean isComplete() {
                 return destination != null && fileUpload.renameTo(destination);
             } catch (IOException e) {
                 throw new MultipartException("Error transferring file: " + fileUpload.getName(), e);
-            } finally {
-                fileUpload.release();
             }
         };
-        if (isComplete()) {
-            return new AsyncSingleResultPublisher<>(ioExecutor, transferOperation);
-        } else {
-            return Observable.<Boolean>create((emitter) -> {
-
-                subject.subscribeOn(Schedulers.from(ioExecutor))
-                    .subscribe(Functions.emptyConsumer(),
-                        (t) -> emitter.onError((Throwable) t),
-                        () -> {
-                            if (fileUpload.isCompleted()) {
-                                emitter.onNext(transferOperation.get());
-                                emitter.onComplete();
-                            } else {
-                                emitter.onError(new MultipartException("Transfer did not complete"));
-                            }
-                        });
-
-            }).firstOrError().toFlowable();
-        }
+
+        return Observable.<Boolean>create((emitter) -> {
+
+            subject.subscribeOn(Schedulers.from(ioExecutor))
+                .subscribe(new Subscriber() {
+                    Subscription subscription;
+                    @Override
+                    public void onSubscribe(Subscription s) {
+                        subscription = s;
+                        subscription.request(1);
+                    }
+
+                    @Override
+                    public void onNext(Object o) {
+                        subscription.request(1);
+                    }
+
+                    @Override
+                    public void onError(Throwable t) {
+                        emitter.onError(t);
+                    }
+
+                    @Override
+                    public void onComplete() {
+                        if (fileUpload.isCompleted()) {
+                            emitter.onNext(transferOperation.get());
+                            emitter.onComplete();
+                        } else {
+                            emitter.onError(new MultipartException("Transfer did not complete"));
+                        }
+                    }
+                });
+        }).firstOrError().toFlowable();
```

Comment:
Does it make sense to use `Single.create` instead of `Observable` here rather than call `firstOrError()`?

Reply:
I didn't change that code as part of this PR, but sure I can update this.